fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/fluid/FluidStorageUtil.java

```diff
@@ -0,0 +1,96 @@
+import net.minecraft.sound.SoundEvent;
+import net.minecraft.util.Hand;
+
+import net.fabricmc.fabric.api.transfer.v1.context.ContainerItemContext;
+import net.fabricmc.fabric.api.transfer.v1.storage.Storage;
+import net.fabricmc.fabric.api.transfer.v1.storage.StorageView;
+import net.fabricmc.fabric.api.transfer.v1.transaction.Transaction;
+
+/**
+ * Helper functions to work with fluid storages.
+ *
+ * <p><b>Experimental feature</b>, we reserve the right to remove or change it without further notice.
+ * The transfer API is a complex addition, and we want to be able to correct possible design mistakes.
+ */
+@ApiStatus.Experimental
+public final class FluidStorageUtil {
+	/**
+	 * Try to make the item in a player hand "interact" with a fluid storage.
+	 * This can be used when a player right-clicks a tank, for example.
+	 *
+	 * <p>More specifically, this function tries to find a fluid storing item in the player's hand.
+	 * Then, it tries to fill that item from the storage. If that fails, it tries to fill the storage from that item.
+	 *
+	 * <p>Only up to one fluid variant will be moved, and the corresponding emptying/filling sound will be played.
+	 * In creative mode, the player's inventory will not be modified.
+	 *
+	 * @param storage The storage that the player is interacting with.
+	 * @param player The player.
+	 * @param hand The hand that the player used.
+	 * @return True if some fluid was moved.
+	 */
+	public static boolean interactWithFluidStorage(Storage<FluidVariant> storage, PlayerEntity player, Hand hand) {
+		// Check if hand is a fluid container.
+		Storage<FluidVariant> handStorage = ContainerItemContext.forPlayerInteraction(player, hand).find(FluidStorage.ITEM);
+		if (handStorage == null) return false;
+
+		// Try to fill hand first, otherwise try to empty it.
+		return moveWithSound(storage, handStorage, player, true) || moveWithSound(handStorage, storage, player, false);
+	}
+
+	private static boolean moveWithSound(Storage<FluidVariant> from, Storage<FluidVariant> to, PlayerEntity player, boolean fill) {
+		for (StorageView<FluidVariant> view : from) {
+			if (view.isResourceBlank()) continue;
+			FluidVariant resource = view.getResource();
+			long maxExtracted;
+
+			// check how much can be extracted
+			try (Transaction extractionTestTransaction = Transaction.openOuter()) {
+				maxExtracted = view.extract(resource, Long.MAX_VALUE, extractionTestTransaction);
+				extractionTestTransaction.abort();
+			}
+
+			try (Transaction transferTransaction = Transaction.openOuter()) {
+				// check how much can be inserted
+				long accepted = to.insert(resource, maxExtracted, transferTransaction);
+
+				// extract it, or rollback if the amounts don't match
+				if (accepted > 0 && view.extract(resource, accepted, transferTransaction) == accepted) {
+					transferTransaction.commit();
+
+					SoundEvent sound = fill ? FluidVariantAttributes.getFillSound(resource) : FluidVariantAttributes.getEmptySound(resource);
```

Comment:
Hmm, this doesn't take into account that the item might have sounds of its own (in vanilla for example, bottles have different sounds than buckets when you interact with cauldrons)

Reply:
I don't think there is much that we can do about this. Imo it is a "cosmetic" oversight (i.e. it's not essential), and addressing it would require too much effort.

Reply:
I don't think it would be too much - or barely any - effort to add two `FluidVariantAttribute` methods (fill/empty sound) that accept an `ItemStack` and default to the non-item stack overloads.

Reply:
Well, you'd expect a modded bottle of some kind to make the same sound as a vanilla bottle, so you'd need some way to group sounds together (I think).

Reply:
The only way to solve that is an event design, but working for vanilla bottles at the very least would be good.

IMO interacting with a modded tank should be able to sound the same as interacting with cauldrons without having to [hardcode bottles](https://github.com/Juuxel/Adorn/blob/2aaf03632b5b1679a8b6f0dce76e658ab107856e/common/src/main/kotlin/juuxel/adorn/block/entity/KitchenSinkBlockEntity.kt#L69-L71).

Reply:
Hmmm yeah, hardcoding bottles is not ideal.

Reply:
Hardcoding bottles for now until a better solution is designed.